File: vuln-sandbox/ruby/cwe-918/ssrf-safe.rb
# CWE-918: Server-Side Request Forgery - Safe Pattern
# Rule: ruby-ssrf-net-http
# Expected: TrueNegative

require 'net/http'
require 'uri'

class WebhookController < ApplicationController
  ALLOWED_HOSTS = %w[api.example.com api.trusted.com].freeze

  def fetch
    # SAFE: Validate URL against allowlist before making request
    uri = URI.parse(params[:url])
    raise "Unauthorized host" unless ALLOWED_HOSTS.include?(uri.host)
    response = Net::HTTP.get(uri)
    render plain: response
  end

  def proxy
    # SAFE: Static URL, no user input
    uri = URI("https://api.example.com/data")
    Net::HTTP.get_response(uri)
  end
end

File: vuln-sandbox/php/cwe-918/ssrf-safe.php
<?php
// CWE-918: Server-Side Request Forgery - Safe Pattern
// Rule: php-ssrf-curl-user-url
// Expected: TrueNegative

$allowed_hosts = ['api.example.com', 'api.trusted.com'];

// SAFE: Validate URL against allowlist before making request
$url = $_GET['url'];
$parsed = parse_url($url);
if (!in_array($parsed['host'] ?? '', $allowed_hosts)) {
    http_response_code(403);
    exit("Unauthorized host");
}

$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, "https://api.example.com/data");
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
$response = curl_exec($ch);
curl_close($ch);
echo htmlspecialchars($response);
?>
